### PR TITLE
Multi-Category support

### PR DESCRIPTION
**Pull request by TheBusyBiscuit**
## Description

This PR adds the `MultiCategory` which itself can store several `SubCategories`.
This will likely not be used inside Slimefun itself. But is relevant for many Addon Creators who wanted a feature like this.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Review by Sfiguz7 — APPROVED** [review]
Oh yes please, some Addons will be able to use this for a much tidier view I love it
As for the code, couldn't spot any issues

**Review by svr333 — APPROVED** [review]
Seems good to me, can't wait to see this used in addons

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+     *            The {@link SubCategory} to remove.
+     */
+    public void removeSubCategory(@Nonnull SubCategory category) {
+        Validate.notNull(category, "The Category cannot be null!");
+
+        subCategories.remove(category);
+    }
+
+    @Override
+    public boolean isVisible(Player p, PlayerProfile profile, SlimefunGuideMode mode) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java:64]:
```suggestion
    @Override
    ParametersAreNonnullByDefault
    public boolean isVisible(Player p, PlayerProfile profile, SlimefunGuideMode mode) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        return true;
+    }
+
+    @Override
+    public void open(Player p, PlayerProfile profile, SlimefunGuideMode mode) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java:69]:
```suggestion
    @Override
    ParametersAreNonnullByDefault
    public void open(Player p, PlayerProfile profile, SlimefunGuideMode mode) {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/SubCategory.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        super(key, item, tier);
+
+        Validate.notNull(parent, "The parent category cannot be null");
+
+        multiCategory = parent;
+        parent.addSubCategory(this);
+    }
+
+    @Override
+    public final boolean isHidden(Player p) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/SubCategory.java:43]: @Nonnull

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/CheatSheetSlimefunGuide.java [hunk 4]
```diff
@@ -73,7 +73,7 @@ public ItemStack getItem() {
     }
 
     @Override
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/CheatSheetSlimefunGuide.java:75]:
```suggestion
    @Override
    @ParamParametersAreNonnullByDefault
```

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @ParametersAreNonnullByDefault
+    private void showBookMenu(Player p, PlayerProfile profile, int page) {
+        profile.getGuideHistory().add(this, page);
+
+        List<ChatComponent> lines = new LinkedList<>();
+
+        for (SubCategory category : subCategories) {
+            ChatComponent chatComponent = new ChatComponent(ChatUtils.crop(ChatColor.DARK_GREEN, ItemUtils.getItemName(category.getItem(p))) + "\n");
```
[the remark this answers is not in the document]
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java]: The Book variant should be removed here anyway...

[the remark this answers is not in the document]
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java]: oh, you reviewed an outdated diff, that's why

[the remark this answers is not in the document]
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/categories/MultiCategory.java]: oh, I had an old review still :lul:

**Comment by Hellcode48** [comment]
I'm super stoked to see this done

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/CheatSheetSlimefunGuide.java (hunk 4 above)
Comment (Sfiguz7) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/guide/CheatSheetSlimefunGuide.java:75]: ParamParameters :p

**Review by Sfiguz7 — APPROVED** [review]
Walshy's annotations to save some technical debt but everything looks good to me

**Review by Sfiguz7 — APPROVED** [review]
LGTM

**Review by WalshyDev — APPROVED** [review]
LGTM

**Pull request merged**